### ISSUE TITLE
Some programs/variables not available

### ISSUE THREAD

**Issue by roberthamilton-pm** [mjcumming/ISY994v5#10]
My ISY can sometimes be slow to return pages like http://<ipaddress>/rest/vars/get/2 - based on this I've been able to improve the reliability of ISY994v5 by changing the http requests to use a 15 second timeout for the the initial requests. Given how important it is to get these definitions during start-up, perhaps it's worth having a different timeout value for initialisation vs 'normal' execution?

Even with the increased timeout, I see some issues where variables and programs aren't found when events arrive via the websocket.

I routinely see e.g. "Unable able to find program type id 00EB" or "Unable able to find variable type 14 id 1" - these definitely exist on the ISY.

Happy to share detailed logs to help debug this.

**Comment by mjcumming** [comment]
Hi, try out the latest version which used the aiohttp library. You can switch the ISY to chunked mode on the settings page from the admin app. This might solve the problem. Let me know.

**Comment by roberthamilton-pm** [comment]
I'm using version 0.7.5, and chunked mode is on.

During startup I see the following error thrown: **HTTP Get Error Cannot write to closing transport**

I routinely see "no variables found of type 2: ('no response for request: ', 'vars/definitions/2')" — so the library can't see any of my state variables :(

**Comment by mjcumming** [comment]
In your home directory should be a isy log file.

**Comment by mjcumming** [comment]
Also what firmware are you using on the isy?

**Comment by mjcumming** [comment]
And … do you have any non state variables declared? Can you add one to see if the problem goes away?

**Comment by roberthamilton-pm** [comment]
ISY is 5.0.16C, and I do have non-state variables too. I'm not seeing a log file generated by the library.

I did a bunch of clean up work today - removing unused variables, deleting many legacy programs, but it hasn't improved things much. One thing I have observed is that if I get any variables, I get all of them - not a subset like before. With programs, the issue remains that quite a few don't get imported from the definitions returned from ISY. I've tried looking for a pattern (e.g. they're in the same subfolder, etc.) but not been able to identify any common behaviours. It is the same set of programs which are skipped over each time, however.

There's an "Invalid scene info" error shown during startup - I'm investigating that now to see if it's connected at all.

Really appreciate your help with this - thank you.

**Comment by mjcumming** [comment]
Are you able to upload the log file?

**Comment by roberthamilton-pm** [comment]
Weirdly I'm not getting a logfile at the moment - is there a switch somewhere to turn on logging?

**Comment by roberthamilton-pm** [comment]
Also weird, the library reports a 404 when trying to access http://192.168.1.9/rest/vars/definitions/2 - I can always get that page in a browser. Not a credentials issue - the library is passing those correctly.

**Comment by roberthamilton-pm** [comment]
One mystery solved: I have some programs which are frequently run but are not enabled. They're valid because the If part of the program evaluates state variables, but they're disabled because I want them to be called only from other programs, not when the value of the state variable changes.

Since program_info.py skips those programs when processing the results from /rest/programs/?subfolders=true, when the programs are subsequently run, the library can't find them and logs an error.

The issue of "missing programs" went away when I commented out:

             if node.attrib["enabled"] == "false":  # not enabled
                return

in program_info.py

I'm still trying to find the root cause for the /rest interface to my ISY being slow and not allowing variable definitions to load, but this is definitely the issue for my "missing" programs.

**Comment by mjcumming** [comment]
I will remove the check for enabled=="false"

Log file should be in your home directory. See controller.py for the logging code.

Are the variables loading fine when you put the url in to a browser?

**Comment by mjcumming** [comment]
Let me know if this is still a problem

**Comment by roberthamilton-pm** [comment]
It still occurs, but I think the root cause is that my ISY sometimes takes several tens of seconds to return a response on the REST interface. I'm now logging this, to try to understand the root cause on the ISY.

Thanks for following up!

**Comment by roberthamilton-pm** [comment]
I implemented lots of logging to try to get to the root cause here... my ISY was often slow to respond to HTTP requests. Fixed that (ISY was repeatedly trying to write changes to battery-powered devices), but still sometimes have issues getting the variable definitions in ISY994v5 library.

In variable_container.py I set the timeout to 2s for the requests and have since had no issues. Not sure what the default HTTP timeout is - I can find out if it's helpful.
